### PR TITLE
feat: Undo and redo for splines

### PR DESCRIPTION
**Pull request by SilviaZeta**
# Description
closes #gliff-ai/annotate/issues/252

# Dependency changes
No

# Testing
No

# Documentation
No

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] New migrations have been committed
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by ChasNelson1990** [comment]
@SilviaZeta have you seen the LGTM errors?

File: src/annotation/index.ts [hunk 1]
```diff
@@ -276,31 +290,59 @@ export class Annotations {
...
+      ]);
+    }
   }
 
   @log
-  insertSplinePoint(idx: number, point: XYPoint): void {
+  insertSplinePoint(idx: number, point: XYPoint, addToUndoRedo = true): void {
     this.data[this.activeAnnotationID].spline.coordinates.splice(idx, 0, point);
+    if (addToUndoRedo) {
+      this.updateUndoRedoActions("deleteSplinePoint", [idx]);
```
Comment (philipjackson) [inline src/annotation/index.ts:344]: Doesn't this also need an `this.updateUndoRedoActions("addSplinePoint", ... `?

**Comment by philipjackson** [comment]
Also I think redo needs to be cleared whenever the user applies an action that's not Redo. E.g. if I add five spline points, undo the last and then add it again in a different place, clicking Redo should then do nothing; right now it'll re-add the undone point. At least that's how undo/redo would be expected to work I think.

**Comment by philipjackson** [comment]
One thing you could change would be to add a single function with a big `switch`/`if`/`else if` chain that inverts actions, that way you don't have to do the inverting in the action methods themselves, and then you could do the undo logging with a decorator instead of having to modify each method, and you could clear the redo array in that decorator instead of copying that step in each method. Doubt it would save much code redundancy overall, but it would be neat (maybe for another PR).

**Comment by philipjackson** [comment]
Two more things:
1. Undoing spline point dragging goes one tiny movement at a time - this should probably be fixed in SplineCanvas, which could make a single call to `updateSplinePoint` in `onMouseUp` instead of a separate call in each `onMouseMove`. Separate PR though, I'll make an issue for it.
2. @ChasNelson1990 , might be worth noting that right now the undo/redo actions themselves are not logged, rather, if you add a spline point then undo it, the undo will be logged in the same way as if you'd hit delete, so the audit won't technically say if an action was caused by hitting undo/redo or doing it manually. This might be fixable but would probably be a bit of a pain

**Comment by ChasNelson1990** [comment]
> 2\. @ChasNelson1990 , might be worth noting that right now the undo/redo actions themselves are not logged, rather, if you add a spline point then undo it, the undo will be logged in the same way as if you'd hit delete, so the audit won't technically say if an action was caused by hitting undo/redo or doing it manually. This might be fixable but would probably be a bit of a pain

I think it's a pain that we should do though, otherwise you won't actually be able to recreate the annotation from the audit if undo has been used.

**Comment by philipjackson** [comment]
> > 2. @ChasNelson1990 , might be worth noting that right now the undo/redo actions themselves are not logged, rather, if you add a spline point then undo it, the undo will be logged in the same way as if you'd hit delete, so the audit won't technically say if an action was caused by hitting undo/redo or doing it manually. This might be fixable but would probably be a bit of a pain
> 
> I think it's a pain that we should do though, otherwise you won't actually be able to recreate the annotation from the audit if undo has been used.

You should be able to re-create it from audit as it is, no? The undo/redo actions are logged, it's just that they're logged as e.g. "delete spline point" rather than "undo add spline point".

**Review by philipjackson — DISMISSED** [review]

**Comment by ChasNelson1990** [comment]
@joshuajames-smith any thoughts on the placement of undo redo?
(Not sure there's a better location but would be good to get your input and icons)

**Comment by joshuajames-smith** [comment]
> @joshuajames-smith any thoughts on the placement of undo redo?
> (Not sure there's a better location but would be good to get your input and icons)

Had a quick think about location, and I think for now it could get away with living in the spline tool bar however I think the UX might be a bit clunky so I think we maybe have two options:
1. We can put the undo redo within the spline tool bar. However, atm the extra tool bars close when the user clicks off tool bar, this may cause some UX hurdles for uses. They would have to click spline > click spline tool > annotate and then to undo they would have to once again click spline > click undo. Maybe the tool bars need to remain open to not slow down the user?
Spline Styling –  tool bar

or
2. Could the undo redo be used across all tools? In which case, could it live in the tool bar which hosts the 'add/delete annotation' etc.?

**Comment by philipjackson** [comment]
> 2. Could the undo redo be used across all tools? In which case, could it live in the tool bar which hosts the 'add/delete annotation' etc.?

This definitely makes more sense to me, the only reason undo/redo are currently associated with the spline tool is that we haven't gotten around to implementing them for the other tools, right? I think they should be in the 'add/delete annotations' toolbar and should be permanently visible, given that they're generic tools that are likely to be used often.

**Review by ChasNelson1990 — APPROVED** [review]

**Pull request merged**